src/lock_free/mwcas_descriptor.cpp

```diff
@@ -0,0 +1,167 @@
+{
+/*##############################################################################
+ * Static utilities
+ *############################################################################*/
+
+auto
+MwCASDescriptor::GetDescriptor()  //
+    -> MwCASDescriptor *
+{
+  /*auto *page = _gc->GetPageIfPossible<MwCASDescriptor>();
+  auto *desc = (page == nullptr) ? new MwCASDescriptor{} : static_cast<MwCASDescriptor *>(page);
+  desc->target_cnt_ = 0;*/
+  // GCを無視した一時的な実装
+  return new MwCASDescriptor{};
+}
+
+/*##############################################################################
+ * Utilities
+ *############################################################################*/
+auto
+MwCASDescriptor::MwCAS()  //
+    -> bool
+{
+  // set a memory fence
+  stat_.store(kUndecided, kRelease);
+  const auto succeeded = MwCASInternal();
+  return succeeded;
+}
+
+//                wordのビット割り当て（実装後に消す）
+// |     63     |       62-50       |      49-47     |        46-0       |
+// | MwCAS Flag | Reference Counter | Begin Position |Descriptor Address |
+
+auto
+MwCASDescriptor::MwCASInternal(  //
+    const size_t begin_pos       //
+    ) -> bool
+{
+  const auto desc_addr = std::bit_cast<uint64_t>(this) | kMwCASFlag;
+  auto cur_stat = stat_.load(kSeqCst);
+  if (cur_stat == kUndecided) {
+    auto stat = kSucceeded;
+    for (size_t i = begin_pos; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      if (!EmbedDescriptor(desc_addr, i)) {
+        stat = kFailed;
+        break;
+      }
+    }
+    cur_stat = stat_.load(kSeqCst);
+    if (cur_stat == kUndecided && stat_.compare_exchange_strong(cur_stat, stat, kSeqCst, kSeqCst)) {
+      cur_stat = stat;
+    }
+  }
+
+  auto ret = (cur_stat == kSucceeded);
+  auto target_cnt_sum = 0;
+  if (ret) {
+    for (size_t i = 0; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      auto expected = target.addr->load(kSeqCst);
+      if (((expected ^ desc_addr) & kDescMask) == 0UL) {
+        target.addr->compare_exchange_strong(expected, target.new_val, target.fence, target.fence);
+      }
+      target_cnt_sum += target.cnt;
+    }
+  } else {
+    for (size_t i = 0; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      auto expected = target.addr->load(kSeqCst);
+      if (((expected ^ desc_addr) & kDescMask) == 0UL) {
+        target.addr->compare_exchange_strong(expected, target.old_val, target.fence, target.fence);
+      }
+      target_cnt_sum += target.cnt;
+    }
+  }
+
+  if (target_cnt_sum) {  // slow path
+    // increment 'exit_cnt_' atomically
+    while (true) {  // 1加算するだけだからこうしたけど，もしかしたら無限ループになるかも
+      auto exit_cnt_now = exit_cnt_.load(kSeqCst);
+      if (exit_cnt_.compare_exchange_weak(exit_cnt_now, exit_cnt_now + 1, kSeqCst, kSeqCst)) {
+        if (exit_cnt_now == (target_cnt_sum + 1)) {
+          // GC処理（未実装）
+        }
+        break;
+      }
+      CPP_UTILITY_SPINLOCK_HINT
+    }
+  } else {  // fast path
+    // GC処理（未実装）
+  }
```

Comment:
ここの加算については [std::atomic::fetch_add](https://cpprefjp.github.io/reference/atomic/atomic/fetch_add.html) が使えます．与えた値を不可分に加算し，加算前の値を返してくれる関数です．

なお，ちょっとしたポイントとして，追従処理に参加したスレッドの数とexitしたスレッドの数が1つずれる（もともとの所有者は追従スレッドとしてカウントされない）という点があります．これにより， `fetch_add` で返ってきた数を直に追従スレッドの総数と比較できます．

```suggestion
  if (target_cnt_sum == 0 || exit_cnt_.fetch_add(1, kSeqCst) == follow_cnt) {
    // GC処理
  }
```

あと，しれっと書き換えてますが `target_cnt_sum` よりは `follow_cnt` とかの方が意図がわかりやすい気がします．